enterprise/frontend/src/embedding-sdk/hooks/private/use-load-question.ts

```diff
@@ -0,0 +1,123 @@
+import { useRef, useState } from "react";
+import { useAsyncFn, useUnmount } from "react-use";
+
+import {
+  runQuestionOnLoadSdk,
+  runQuestionOnQueryChangeSdk,
+  runQuestionOnNavigateSdk,
+} from "embedding-sdk/lib/interactive-question";
+import type {
+  LoadSdkQuestionParams,
+  NavigateToNewCardParams,
+  SdkQuestionResult,
+} from "embedding-sdk/types/question";
+import { defer, type Deferred } from "metabase/lib/promise";
+import { useDispatch } from "metabase/lib/redux";
+import type Question from "metabase-lib/v1/Question";
+
+export interface LoadQuestionHookResult {
+  question?: Question;
+  queryResults?: any[];
+
+  isQuestionLoading: boolean;
+  isQueryRunning: boolean;
+
+  loadQuestion(): Promise<SdkQuestionResult & { originalQuestion?: Question }>;
+  onQuestionChange(question: Question): Promise<void>;
+  onNavigateToNewCard(params: NavigateToNewCardParams): Promise<void>;
+}
+
+export function useLoadQuestion({
+  location,
+  params,
+}: LoadSdkQuestionParams): LoadQuestionHookResult {
+  const dispatch = useDispatch();
+
+  // Keep track of the latest question and query results.
+  // They can be updated from the below actions.
+  const [result, setQuestionResult] = useState<SdkQuestionResult>({});
+  const { question, queryResults } = result;
```

Comment:
We need a fresh `queryResult` as well as a fresh `question` to use in `previousQuestion` which can comes from each of the 3 actions, so we can't simply use `loadQuestionState.value`, `questionChangeState.value` and `navigateToNewCardState.value` as the source of truth.

This is because we wouldn't know which action has the freshest question and query result, as each actions could've been called anytime. (e.g. `loadQuestion` can be called when the question is reset)